### PR TITLE
[ PR-1071-2] add sendRequest.done() to release resource together

### PR DESCRIPTION
**Pull request by gunli**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #1071

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #1071

### Motivation

Spilt PR #1071 into multiple ones, this is the SECOND one.

https://github.com/apache/pulsar-client-go/pull/1071 is a BIG PR, it refactor the sending logic, here we just add fields to `sendRequest` add `sendRequest.done()`  to make `sendRequest` can keep track of the resouces it applied and can release them together.
### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / GoDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by gunli** [comment]
@tisonkun @RobertIndie @Gleiphir2769 @nodece

**Comment by tisonkun** [comment]
@nodece If you apporve this patch to be merged, please give a formal review reaction :D

**Review by tisonkun — APPROVED** [review]
Let's go forward. Although, I don't like this spilt flavor that we add methods or fields that is unused in the patch - each patch should be self-contained and we add methods or fields when we do use it.

Given #1071 pending for a long time and I review the following PR that use these fields and methods, I can merge this one. But if it's possible please try to make patches self-contained instead of adding something unused first and letting the reviewers guess how it will be used.

File: pulsar/producer_partition.go [hunk 1]
```diff
@@ -1419,21 +1419,37 @@ func (p *partitionProducer) Close() {
 	<-cp.doneCh
 }
 
+//nolint:all
```
Comment (tisonkun) [inline pulsar/producer_partition.go:1422]: Ensure we remove this `nolint` in a following patch.

File: pulsar/producer_partition.go [hunk 2]
```diff
@@ -1443,6 +1459,55 @@ func (sr *sendRequest) stopBlock() {
 	})
 }
 
+//nolint:all
```
Comment (tisonkun) [inline pulsar/producer_partition.go:1462]: ditto

**Pull request merged**